source/h2_decoder.c

```diff
@@ -627,21 +718,52 @@ static int s_state_fn_frame_settings(struct aws_h2_decoder *decoder, struct aws_
     return AWS_OP_SUCCESS;
 }
 static int s_state_fn_frame_push_promise(struct aws_h2_decoder *decoder, struct aws_byte_cursor *input) {
-    (void)decoder;
-    (void)input;
-    /* #TODO: make go */
-    return aws_raise_error(AWS_ERROR_UNIMPLEMENTED);
+
+    AWS_FATAL_ASSERT(input->len >= 4);
+    AWS_FATAL_ASSERT(decoder->next == NULL);
+
+    uint32_t promised_stream_id = 0;
+    bool succ = aws_byte_cursor_read_be32(input, &promised_stream_id);
+    AWS_FATAL_ASSERT(succ);
+    decoder->frame_in_progress.payload_len -= 4;
+
+    /* Remove top bit */
+    promised_stream_id &= s_31_bit_mask;
+
+    /* Alert the user */
+    DECODER_CALL_VTABLE_STREAM_ARGS(decoder, on_push_promise, promised_stream_id);
+
+    /* Update the stream id of the decoder so that future on_header and on_end_headers calls use the promised id */
+    decoder->frame_in_progress.stream_id = promised_stream_id;
+
+    const struct decoder_state *padding = &s_state_padding;
+
+    if (decoder->frame_in_progress.payload_len) {
+        /* Read the headers block */
+        s_decoder_set_state_and_next(decoder, &s_state_headers_begin, padding);
```

Comment:
don't padded if isn't padded
paddeded

actually always padded because catching this every time is too tricky for our human brains